### PR TITLE
IIO Trigger now support with example on EVAL-ADXL355-PMDZ

### PR DESCRIPTION
**Pull request by rbolboac**
- drivers:accel:adxl355 Moddified return value for error_com label 
- Added doxy comment for is_synchronous field from iio_trigger structure.
- Added null pointer check in iio_adxl355_trigger_init for init_param->iio_desc and init_param->name.
- Added IIO Trigger now support with example on EVAL-ADXL355-PMDZ

### CODE REVIEW

File: iio/iio.c [hunk 1]
```diff
@@ -679,48 +718,69 @@ static int iio_read_attr(struct iiod_ctx *ctx, const char *device,
...
-	    strcmp(attr->name, REG_ACCESS_ATTRIBUTE) == 0) {
-		if (dev->dev_descriptor->debug_reg_read)
-			return debug_reg_read(dev, buf, len);
+	/* If IIO device with given name is found, handle reading of attributes */
+	if (dev) {
+		if (attr->type == IIO_ATTR_TYPE_DEBUG &&
+		    strcmp(attr->name, REG_ACCESS_ATTRIBUTE) == 0) {
+			if (dev->dev_descriptor->debug_reg_read)
+				return debug_reg_read(dev, buf, len);
+			else
```
Comment (amiclaus) [inline iio/iio.c:736]: redundant else.

File: iio/iio.c [hunk 2]
```diff
@@ -679,48 +718,69 @@ static int iio_read_attr(struct iiod_ctx *ctx, const char *device,
...
+		params.buf = buf;
+		params.len = len;
+		params.dev_instance = dev->dev_instance;
+		attributes = get_attributes(attr->type, dev, ch);
+		if (!strcmp(attr->name, ""))
+			return iio_read_all_attr(&params, attributes);
 		else
-			return -ENOENT;
-	}
+			return iio_rd_wr_attribute(&params, attributes, attr->name, 0);
```
Comment (amiclaus) [inline iio/iio.c:763]: redundant else.

File: iio/iio.c [hunk 3]
```diff
@@ -679,48 +718,69 @@ static int iio_read_attr(struct iiod_ctx *ctx, const char *device,
...
+		/* If IIO trigger with given name is found, handle reading of attributes */
+		if(trig_dev) {
+			params.ch_info = NULL; /* Triggers cannot have channels */
+			params.buf = buf;
+			params.len = len;
+			params.dev_instance = trig_dev->instance;
+			attributes = get_trig_attributes(attr->type, trig_dev);
+			if (!strcmp(attr->name, ""))
+				return iio_read_all_attr(&params, attributes);
+			else
```
Comment (amiclaus) [inline iio/iio.c:777]: redundant else.

File: iio/iio.c [hunk 4]
```diff
@@ -679,48 +718,69 @@ static int iio_read_attr(struct iiod_ctx *ctx, const char *device,
...
-		ch_info.ch_out = ch_out;
-		ch_info.ch_num = ch->channel;
-		ch_info.type = ch->ch_type;
-		ch_info.differential = ch->diferential;
-		ch_info.address = ch->address;
-		params.ch_info = &ch_info;
-	} else {
-		params.ch_info = NULL;
+		} else {
+			return -ENODEV;
```
Comment (amiclaus) [inline iio/iio.c:781]: redundant else

File: iio/iio.c [hunk 5]
```diff
@@ -736,49 +796,70 @@ static int iio_write_attr(struct iiod_ctx *ctx, const char *device,
...
-		ch_out = attr->type == IIO_ATTR_TYPE_CH_OUT ? 1 : 0;
-		ch = iio_get_channel(attr->channel, dev->dev_descriptor,
-				     ch_out);
-		if (!ch)
-			return -ENOENT;
+		if (attr->type == IIO_ATTR_TYPE_DEBUG &&
+		    strcmp(attr->name, REG_ACCESS_ATTRIBUTE) == 0) {
+			if (dev->dev_descriptor->debug_reg_write)
+				return debug_reg_write(dev, buf, len);
+			else
```
Comment (amiclaus) [inline iio/iio.c:815]: redundant else

File: iio/iio.c [hunk 6]
```diff
@@ -736,49 +796,70 @@ static int iio_write_attr(struct iiod_ctx *ctx, const char *device,
...
-		return iio_write_all_attr(&params, attributes);
-	else
-		return iio_rd_wr_attribute(&params, attributes, attr->name, 1);
+		params.buf = (char *)buf;
+		params.len = len;
+		params.dev_instance = dev->dev_instance;
+		attributes = get_attributes(attr->type, dev, ch);
+		if (!strcmp(attr->name, ""))
+			return iio_write_all_attr(&params, attributes);
+		else
```
Comment (amiclaus) [inline iio/iio.c:842]: redundant else

File: iio/iio.c [hunk 7]
```diff
@@ -736,49 +796,70 @@ static int iio_write_attr(struct iiod_ctx *ctx, const char *device,
...
+		/* If IIO trigger with given name is found, handle writing of attributes */
+		if(trig_dev) {
+			params.ch_info = NULL; /* Triggers cannot have channels */
+			params.buf = (char *)buf;
+			params.len = len;
+			params.dev_instance = trig_dev->instance;
+			attributes = get_trig_attributes(attr->type, trig_dev);
+			if (!strcmp(attr->name, ""))
+				return iio_read_all_attr(&params, attributes);
+			else
```
Comment (amiclaus) [inline iio/iio.c:856]: redundant else

File: iio/iio.c [hunk 8]
```diff
@@ -736,49 +796,70 @@ static int iio_write_attr(struct iiod_ctx *ctx, const char *device,
...
+				return iio_rd_wr_attribute(&params, attributes, attr->name, 1);
+
+		} else {
+			return -ENODEV;
```
Comment (amiclaus) [inline iio/iio.c:860]: redundant else

**Review by amiclaus — COMMENTED** [review]
lgtm, just some minor comments. Also for the commit where only comments are added, it would be nice to specify in the commit description "No functional changes" (but not mandatory)

File: iio/iio.c (hunk 6 above)
Comment (rbolboac) [reply iio/iio.c:842]: could you please elaborate? this else is actually being used to call iio_rd_wr_attribute to write to a single attribute.

File: iio/iio.c (hunk 7 above)
Comment (rbolboac) [reply iio/iio.c:856]: this else is actually being used to call iio_rd_wr_attribute to read a single attribute.

**Comment by rbolboac** [comment]
Implemented review changes:
- removed redundant else
- changed commit message to contain: "No functional changes".

**Review by amiclaus — APPROVED** [review]

File: drivers/accel/adxl355/iio_adxl355_trig.c [hunk 9]
```diff
@@ -149,6 +195,45 @@ int iio_adxl355_trigger_init(struct adxl355_iio_trig **iio_trig,
...
+	if (!init_param->iio_desc) {
+		ret =-EINVAL;
+		goto error;
+	}
+	trig_desc->iio_desc = init_param->iio_desc;
+
+	if(!init_param->name) {
+		ret=-EINVAL;
+		goto error;
+	}
```
Comment (CiprianRegus) [inline drivers/accel/adxl355/iio_adxl355_trig.c:225]: These 2 checks can be made before `trig_desc` is allocated, which will simplify the code a bit. The same applies for the `iio_adxl355_trigger_init` as well.

File: drivers/accel/adxl355/iio_adxl355_trig.c [hunk 10]
```diff
@@ -149,6 +195,45 @@ int iio_adxl355_trigger_init(struct adxl355_iio_trig **iio_trig,
...
+	strncpy(trig_desc->name, init_param->name, TRIG_MAX_NAME_SIZE);
+
+
```
Comment (CiprianRegus) [inline drivers/accel/adxl355/iio_adxl355_trig.c:228]: nitpick: extra line

**Comment by rbolboac** [comment]
Implemented review changes:
- refactor  iio_adxl355_trigger_init and iio_adxl355_software_trigger_init

File: drivers/accel/adxl355/iio_adxl355_trig.c [hunk 11]
```diff
@@ -149,6 +190,33 @@ int iio_adxl355_trigger_init(struct adxl355_iio_trig **iio_trig,
...
+int iio_adxl355_software_trigger_init(struct adxl355_iio_trig **iio_trig,
+				      struct adxl355_iio_sw_trig_init_param *init_param)
+{
+	struct adxl355_iio_trig *trig_desc;
+
+	if (!init_param->iio_desc || !init_param->name)
+		return  -EINVAL;
+
+	trig_desc = (struct adxl355_iio_trig*)calloc(1, sizeof(*trig_desc));
+
```
Comment (CiprianRegus) [inline drivers/accel/adxl355/iio_adxl355_trig.c:210]: This is missing a NULL check, otherwise looks good.

**Review by CiprianRegus — APPROVED** [review]
Looks good!

**Pull request merged**